File: 125.valid-palindrome.js
/*
 * @lc app=leetcode id=125 lang=javascript
 *
 * [125] Valid Palindrome
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
  const sWithoutSpaceAndPunctuation = s.replace(/\W/g, '').toLowerCase();
  let i = 0;
  const length = sWithoutSpaceAndPunctuation.length;
  while (i < length / 2) {
    if (
      sWithoutSpaceAndPunctuation[i] !==
      sWithoutSpaceAndPunctuation[length - 1 - i]
    ) {
      return false;
    }
    i++;
  }
  return true;
};
// @lc code=end
// 476/476 cases passed (64 ms)
// Your runtime beats 99.29% of javascript submissions
// Your memory usage beats 100 % of javascript submissions (36.8 MB)

File: 125.valid-palindrome.java
/*
 * @lc app=leetcode id=125 lang=java
 *
 * [125] Valid Palindrome
 */

// @lc code=start
class Solution {
    public boolean isPalindrome(String s) {
        int head = 0;
        int tail = s.length() - 1;

        while (head < tail) {
            char sHead = s.charAt(head);
            char sTail = s.charAt(tail);
            if (!Character.isLetterOrDigit(sHead)) {
                head++;
            } else if (!Character.isLetterOrDigit(sTail)) {
                tail--;
            } else {
                if (Character.toLowerCase(sHead) != Character.toLowerCase(sTail)) {
                    return false;
                }
                head++;
                tail--;
            }
        }
        return true;
    }
}
// @lc code=end
// 476/476 cases passed (2 ms)
// Your runtime beats 99.5 % of java submissions
// Your memory usage beats 71.43 % of java submissions (38.8 MB)